airflow/container_runtime.go

```diff
@@ -0,0 +1,113 @@
+package airflow
+
+import (
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
+
+	"github.com/astronomer/astro-cli/pkg/util"
+
+	"github.com/astronomer/astro-cli/config"
+	"github.com/astronomer/astro-cli/pkg/fileutil"
+	"github.com/pkg/errors"
+)
+
+// FileChecker interface defines a method to check if a file exists.
+// This is here mostly for testing purposes. This allows us to mock
+// around actually checking for binaries on a live system as that
+// would create inconsistencies across developer machines when
+// working with the unit tests.
+type FileChecker interface {
+	Exists(path string) bool
+}
+
+// OSFileChecker is a concrete implementation of FileChecker.
+type OSFileChecker struct{}
+
+// Exists checks if the file exists in the file system.
+func (f OSFileChecker) Exists(path string) bool {
+	exists, _ := fileutil.Exists(path, nil)
+	return exists
+}
+
+// FindBinary searches for the specified binary name in the provided PATH directories,
+// using the provided FileChecker. It searches each specific path within the systems
+// $PATH environment variable for the binary concurrently and returns a boolean result
+// indicating if the binary was found or not.
+func FindBinary(pathEnv, binaryName string, checker FileChecker) bool {
+	// Split the $PATH variable into it's individual path,
+	// using the OS specific path separator character.
+	paths := strings.Split(pathEnv, string(os.PathListSeparator))
+
+	// Although programs can be called without the .exe extension,
+	// we need to append it here when searching the file system.
+	if isWindows() {
+		binaryName += ".exe"
+	}
+
+	var wg sync.WaitGroup
+	found := make(chan string, 1)
+
+	// Search each individual path concurrently.
+	for _, dir := range paths {
+		wg.Add(1)
+		go func(dir string) {
+			defer wg.Done()
+			binaryPath := filepath.Join(dir, binaryName)
+			if exists := checker.Exists(binaryPath); exists {
+				found <- binaryPath
```

Comment:
If it's found in multiple paths will this block on the non-first paths because the channel is full, and so `wg.Wait()` will also block?

Reply:
Yep, good eye. I was able to confirm this locally. Fixed this so that after the first found binary path gets sent into the channel, any subsequent binaries found will just be skipped over. All we care about is that there's at least one, so skipping these should be fine. Once they've all finished we just check if we can pull the one item out or not, determining the return value of this function.